New low level

dneise opened the pull request:
## Do not merge, this is just for discussing stuff.

I branched this off, of Max new_api branch.
In this branch, I try to show, what I think the lowest level of this package should do.
It should take the file apart into a list (or otherwise iterable) of these EventIOObjects or however one wants to call them. 

In order to do this, it needs to know about these headers ... you know... these object headers. But it does not need to know about the content. 

Creating just a list, as I did it here, is certainly not smart, since it slubbers the entire file into memory. 
I have no idea what is smarter, the two things springing to mind are:

### A: Make a complete file index:
Just look at all headers and omit looking at the content/payload/data/whatever at all. The resulting list should always fit into memory and the headers contain something like jump addresses, so later code can directly jump to the right place in the file. 

### B: Be ready to make a file index:
Implement a `__getitem__` and just when the user wants us to, we skip forward to the Object in question and return the header and the payload together. This can for example be given to some code, that is able to understand what to do with them.

dneise commented:
Next up, is implementing some tests.

dneise commented:
Okay, so that's it for the moment, I would like to hear your opinion about this. 

I like the yield_objects function better, because I can give pure data to the parsers, which completely decouples them from something like a file. All the parsers see is header(s) and binary payload, they have to work on. Nothing like a file ... nothing like seek and tell and stuff. just plain parsing, nicely decoupled and possibly testable. 

The list of headers is not enough for the parsers, they also need an open file handle or a path. Should each parser open the file for himself, seek to the start of his payload and start parsing? It's pretty unclear to me, so I don't like this variant.

dneise commented:
The one test, which is failing, is intentional. The `ObjectHeader` constructs itself from a byte stream only and cannot know at which level it is. It knows if its toplevel or not because thats a boolean it gets as a parameter during construction time. The level is identical to the depth the ObjectHeader has inside the list of lists or the first element of the result of the yield_objects function. 

I don't remember why the headers know about their level, so I left the test alone, so we are aware there is now a feature missing.

dneise commented:
Okay I like this low level API now. What do you think?

MaxNoe commented on eventio/event_io_file.py:
```diff
@@ -2,52 +2,64 @@
+            yield o
+
+def object_list(path):
+    f = open(path, 'rb')
```
I think you should give it a file object. Like this we keep compatibility with all supported compression formats of python (gzip, bzip2 etc)

dneise replied:
I always get this wrong ... thanks.

MaxNoe commented on eventio/base.py:
```diff
@@ -173,112 +170,13 @@ def __repr__(self):
         position = eventio_file_or_object.tell()
         try:
-            header = ObjectHeader(
+            header = make_ObjectHeader(
```
I really do not like this make_ObjectHeader. Looks very unpythonic to me.

dneise replied:
I think you said this somewhere already ... just made it a `ObjectHeader.from_file` classmethod

MaxNoe commented on eventio/event_io_file.py:
```diff
@@ -0,0 +1,65 @@
+    def __getattr__(self, attr):
+        if attr == "payload":
+            if self._file.closed:
+                self._file = open(self._file.name, 'rb')
```
This breaks compatibility with every other file like object on could perfectly use for the first instantiation.

dneise replied:
Right, any idea how to solve it?
The file could be closed before anybody ever tries to access an EventIOObject.payload.
Then we get an exception here.

dneise replied:
Ah, I see your next comment ... The end user is responsible. The `open(path)` is pushed way up to the user, who should not close the file, while still working with `EventIOObjects`, which might want to fetch `payload` from that file. cool

MaxNoe replied:
Yup. If every object keeps a reference to the file, the file will only be closed if the user explicitly calls `file.close()`.

MaxNoe commented on eventio/event_io_file.py:
```diff
@@ -0,0 +1,65 @@
+from .object_header import make_ObjectHeader
+from io import BytesIO
+
+class EventIOObject:
```
Why does this object not know it's own header? this is very surprising I think.

MaxNoe replied:
Butbthisba list of the headers oft its subitems, right? Not the object's own item.

dneise replied:
Sorry?

MaxNoe replied:
The member `EventIOObject.headers` is a list of headers, containing the headers of each subitem of this object, correct?

How do I access the header of the object itself? EventIOObject should either have a member header or variables like `version`, `type`, `length` ...

dneise replied:
Nope. `headers`  is a list of headers belonging to this object. An object has multiple headers. 
As I understand it, an eventio file is a kind of tree. As I walk down from header to header, I ultimately arrive at a leave(payload). 

In order to understand the leave, I need all headers, I collected on the way down to the leave. For example, I need  the topmost header to understand the endiannes of the leave, and the downmost header will tell me its length in bytes.

MaxNoe replied:
Wait, so `self.header` is basically the path down the tree? Wow, that is really unexpected to me.

dneise replied:
I see ... makes sense, since I didn't explain anything anywhere.

I wrote an explanation, but it got too long for a comment here.
So I put it further down.

MaxNoe commented on eventio/event_io_file.py:
```diff
@@ -0,0 +1,65 @@
+    The payload might be loaded lazily on first access or already in memory.
+    This is decided on construction time.
+    '''
+    def __init__(self, headers, payload=None, file=None):
```
These look like two different constructors to me.

I think `EventIObject.from_payload` as a classmethod would be better sutied than an init with two mutual exclusive arguments.

dneise replied:
the classmethod approach as discussed here:
http://stackoverflow.com/q/682504/1532474
has zero upvotes and the author says: "if the constructors do not share code", but we can think about it.

However, I am currently thinking there is reason to always give it the file, and let it load the contents itself. If the user goes linearly through the file, the file pointer is already at the right position, so the `self._file.seek` is basically a no-op, just the calculation of the start_adress costs a bit, since it is not needed, when the pointer is already at the right position. (Who cares?!)

I will try that now ...

dneise replied:
Did it ... remmoved another 16 lines or so ... pretty nice. Thanks

MaxNoe replied:
There is a class methods answer with 350 upvotes http://stackoverflow.com/a/682545/3838691 :)

dneise replied:
cool .. didn't read that far down ... thanks.

MaxNoe commented on eventio/object_header.py:
```diff
@@ -0,0 +1,107 @@
+from builtins import property as _property, tuple as _tuple
```
Why?

dneise replied:
I forgot to delete it

MaxNoe commented on eventio/tests/test_low_level.py:
```diff
@@ -0,0 +1,34 @@
+    'eventio', path.join('resources', '3_gammas_reuse_5.dat')
+)
+
+import eventio.object_header as oh
```
I think the imports should go into the test functions, so a failed import only affects the test where the import is needed

dneise replied:
ah cool ... this also nicely packs them to where they are needed. Thanks.

MaxNoe commented on eventio/event_io_file.py:
```diff
@@ -0,0 +1,47 @@
+
+    def __getattr__(self, attr):
+        if attr == "payload":
+            start_address = sum(h.data_field_first_byte for h in self.headers)
```
Why is start address the sum of the first bytes?

dneise replied:
well spotted, it is not, that was a bug solved in 9a2fbee.

dneise commented:
reply


---------------------------------------------------------------

So from my limited experience I learned there is no fundamental difference between an eventio file and a tar file, both make no assumptions about the payload, i.e. putting 3 encrypted files into a tar-file and saying:
"You should be able to read the data, it is in a tar file, which 
are widely common and easy to use."
Makes no sense. Same is true for eventio.

Anyway before starting to decrypt/interpret the payload, you still need to untar the file and this was the aim of this low level interface. Everything needed to implement it, is taken from "eventio_en.pdf" chapter 2. which I found in my `bernlohr/` folder inside my copy of Corsika. Apart from the naming, which is probably still inconsistent, my bad.

As one can see in eventio_en.pdf chapter 2 figure 1 an eventio file implements a tree like structure. Again similar to the folder structure one can have in tarballs. But unlike a tarball, one can only have payload in the leave-nodes of the tree:

> Bit 30 [of the 'length' field] is set if the object contains only sub-objects but no elementary data types, and therefore its contents can be listed without knowing anything about specific object types.

So either an object contains only sub-objects and no payload on its own, which makes it an inner node or *branch*. Or it does contain unexplained payload, which makes it an external-node or *leave*.

There is no uniqueness constraint at all, so two entirely equal objects could be stored next to each other. So this tree is nothing like a python map. There are no clear unique keys. So I wondered which representation to choose. There are many representations for trees, and to first coming to mind is lists of [tuple(header, list_or_payload), i.e.:

    def EventIOFile(path):
        tree = [
            (H0, P0), 
            (H1, [
                (H1_0, P1_0), 
                (H1_1, P1_1),
            ]), 
            (H2, [
                (H2_0, [
                    (H2_0_0, P2_0_0)
                ]),
            ]),
        ]
        return tree

The low-level interface could return this to the user (or in this case developer on a higher level) and be done. The next thing task, is to decrypt/interpret the payload. In order to interpret a file one needs to know the *file-type*. For interpretation of eventio payloads the *type* is not going to be inferred from the payload itself by means of magic numbers, as is the case for files on disk, so how does one get the *type*? Now as I said earlier, the low level interface is based on "eventio_en.pdf", the interpretation of the payload in contrast is based on some code example we found, which looked highly convoluted to me. So I had to interpret them.

Some interpretations were:

 * If the header has type 1200, then this cannot be a branch, but it must be a leave node. And the payload it has, can be interpreted as a *CorsikaRunHeader*.
 * If a header has type 1204, this cannot be a leave, it must be a branch. All nodes below this branch must be of type 1205.
   If a header has type 1205, this must be a leave. Its payload can be interpreted as IACTPhotonBunches.


From such "rules", given a tree like structure as shown above, one could traverse this tree one way or another, and search for nodes that comply to one of these rules. And if one finds them, one can interpret their payload accordingly. Many tree traversion functions, I've seen are implemented recursively. Recursive functions are not considered the most easy to understand, I think we agree on that. So I wondered if there is a different representation of the eventio-tree, which is maybe more suitable.

I think due to the special kind of this tree, which is inner nodes carry no payload, another simple representation is a list of leave-nodes. In order to create this list, I simply visit all leave nodes, and collect the headers I met on my way down. The result is a complete but 1-dimensional represenation of the eventio-file.
In our example it looks like this:

    def EventIOFile(path):
        tree = [
            ((H0, ), P0),
            ((H1, H1_0) , P1_0),
            ((H1, H1_1) , P1_1)),
            ((H2, H2_0, H2_0_0), P2_0_0),
        ]
        return tree

The advantage, I valued pretty high here, is that each element of this list, is the smallest self contained element, which is enough to allow the interpretation of the payload, I could find. So in order to find out, if an element of this list, is of a certain type. I need to look at the `tuple(h.type for h in headers)`. If this is for example `(1204, 1205)`, then the payload must be IACTPhotonBunches and so one. No need for tree traversing functions in the higher level user code.

dneise commented:
I wanted to see, if this low-level interface also works for the large simtel file:
```
In [1]: path = 'eventio/resources/gamma_test.simtel'

In [2]: import eventio

In [3]: obs = eventio.objects(open(path, 'rb'))

In [4]: def get_types(o):
   ...:     return tuple(h.type for h in o.headers)
   ...: 

In [5]: %time type_set = set(map(get_types, obs))
CPU times: user 36 ms, sys: 0 ns, total: 36 ms
Wall time: 36.6 ms

In [6]: len(type_set)
Out[6]: 223
```

And looking at the type_set:
```
In [7]: type_set
Out[7]: 
{(70, 71),
 (70, 72),
 (1212,),
 (2000,),
 (2001,),
 (2002,),
 (2003,),
 (2004,),
 (2005,),
 (2006,),
 (2007,),
 (2008,),
 (2010, 2009),
 (2010, ...),
 (2010, 2171),
 (2010, 2201, 2011),
 (2010, ...., ....),
 (2010, 2271, 2027),
 (2010, 3102),
 (2010, 3104),
 (2010, 3110),
 (2010, 3117),
 (2010, 3118),
 (2010, 3119),
 (2010, 3124),
 (2010, 3126),
 (2010, 3202, 2011),
...
 (2010, 3204, 2011),
...
 (2010, 3210, 2011),
...
 (2010, 3217, 2011),
 ...
 (2010, 3218, 2011),
...
 (2026,)}
```

So one sees there are 223 possibly different types of payload. And maybe the functions to interpret the payload need to only look at the type of the down most header to understand its type. I have no idea. There is absolutely no documentation about this available to me.

MaxNoe commented:
An eventio type is an enclosed object. The only thing you need to know about it which you cannot get from the header itself (if it is not a top level object) is the endianness, for that you need the topmost object. You never need the objects in between to represent the leaf. Only root object and item itself.

dneise commented:
That's good to know. Is this documented somewhere?

MaxNoe commented:
I interpreted the eventio docs like this. Now that I read it again, it is not specifically mentioned but the schematics kind of indicate that.

MaxNoe commented:
Why did you remove the doc string and `__repr__` of Corsika Event?

MaxNoe commented on eventio/iact/__init__.py:
```diff
@@ -2,49 +2,78 @@
+def generate_event(shower):
+    '''
+    A shower is simply a list of EventIOObject
+    of these kinds:
```
There could be more objects. E.g. PhotoElectrons. Depends on the InputCard. The other implemantation could handle this. This cannot.

MaxNoe replied:
I will try to create one. If I understood the manual correctly, PhotoElectron bunches are written if the `CEFFIC` option is used.

dneise commented:
> Why did you remove the doc string and `__repr__` of Corsika Event?

`__repr__` I removed, because custom representations can lie. I this case the thing is a namedtuple with a lot of members, But this version of `__repr__` was hiding some of them.

The docstring I removed, because it might lie. Once one changes the units in `iact/objects.py` from crazy corsika cm;ns to sane m;s one will for sure forget to update that comment in `iact/__init__.py`.

I have to look into some of these physical units packages and learn if there is a nice way to attach the units to the event already in `iact/objects.py`. Then the user has the units, without us needing to take care of possibly outdated docstrings.

dneise commented:
Ah yes, I should add. 

This PR is about the low_level interface. With that I mean everything inside 
 - `eventio/__init__.py`
 - `eventio/object_header.py`
 - `evention/event_io_file.py`

All the rest, can be seen as examples showing how this change in the low_level interface affects higher level code, as e.g. the code in the `iact/`folder. It should show, that stuff still works, but details like doc strings or reprs are beyond this PR.

MaxNoe commented:
I thought about this tree representation and came to the conclusion I do not really like it. I think we should keep the tree structure and not reduce it to end points. One reason for this is that simtel files seem to have huge substructures. If we keep the tree, one could decide to only load the top-level first and then dive deeper into the tree if one wants to read this top-level object.

dneise commented:
I also looked at the simtel file and came to the same conclusion, which is:
The tree has meaning, even though none of the nodes really have meaning.

I must say, I lost a lot of time chasing ghosts, when using that simtel file, since it is broken.
I thought my code was broken. Imho, we should not use broken files for test cases (which we don't ... but I understood Max N, that he is using that file for testing. So I did the same.)

Unless we define some features like: We want our code to behave like this and that when seeing a truncated file. Then it makes total sense to have a truncated file and use it for testing. 

So I think, implicitly the current master, does have such a feature. But it was neither expressed somewhere, nor exists a test documenting this ... not fun at all. 

---

Anyway, I have written something which returns a different representation of the tree. Namely this list of tuples, I talked about higher up. It is in a different branch. I will open a PR there as well and we can discuss about it there.

MaxNoe commented:
The nodes have meaning, through their types. Even though they are just containers, bernlohr uses different types to tell the user, what's inside the container. One could simply have one container type and give it metadata, but this is not how it is implemented.

For the simtel file: I should have mentioned that is is truncated, sorry. It is the only simtel file I have access to which is not crazily big. When I have the time (and as soon as I really start working on supporting simtel) I will produce small test files.

dneise commented:
> The nodes have meaning, through their types. Even though they are just containers, bernlohr uses different types to tell the user, what's inside the container. One could simply have one container type and give it metadata, but this is not how it is implemented.

Hm ... interesting. Of course you are right, they have (type, id, funny_user_bit) all this contains some information. I think `id` is used by bernloehr to encode something like `event_number` sometimes. So instead of opening a shower and reading the event header in order to know the `event_number` one can use the `id` of the container-like event header, which comes before the shower.

In chapter 2 of "eventio_en.pdf" he writes:
> The ID field should identify from where the data come, which event they represent, or some-
thing of that kind. Thus, the ID field has an object-specific meaning. An ID of −1 should be
interpreted as ‘no ID given.’

Which I translated into:

One can only know, what some nodes meaning is, once one has actually decrypted the the leaves below that node. But then one already knows everything, so ... well what is the point in that?

However, when the **type** of an inner node already explains when its `id` and `funny_user_bit` mean, without the need to go down to the leave, then its useful. 

Cool cool

MaxNoe commented:
From the source code of `mc_tel.h`:
> /* I/O item types: */

/* Never change the following numbers after MC data is created: */
#define IO_TYPE_MC_BASE     1200
#define IO_TYPE_MC_RUNH     (IO_TYPE_MC_BASE+0)
.

Also from the source code (io_simtel.c), the meaning of the id in IACTPhotons is  (1000*reuse + telescope id). The meaning if the id in TelescopeData is the reuse (or array number like he calls it).
